### PR TITLE
use 'final' modifier for fields where possible

### PR DESCRIPTION
**Pull request by turbanoff**
Also removed unused fields which my IDEA highlights as possible `final`, but fields were unused (leftovers after refactoring):
1. `DefaultLogBuilder.EMPTY_MESSAGE`
2. `DirectWriteRolloverStrategy.usePrevTime`
3. `ParameterFormatter.threadLocalSimpleDateFormat`
4. `PropertiesUtil.SYSTEM`
5. `FileAppender.DEFAULT_BUFFER_SIZE`
6. `RollingFileAppender.DEFAULT_BUFFER_SIZE`

### CODE REVIEW

**Comment by turbanoff** [comment]
I didn't touch fields in samples, tests and perf modules.
Field `org.apache.logging.log4j.core.layout.AbstractStringLayout#charset` looks strange. Looks like it should be initialized in `readObject` method, but it was removed here - https://github.com/apache/logging-log4j2/commit/afd417ce965fd6100134ac775aef1be98a927250
See https://github.com/apache/logging-log4j2/commit/afd417ce965fd6100134ac775aef1be98a927250#diff-1c90b17f7481c4e5ae5a134261a36209

File: log4j-1.2-api/src/main/java/org/apache/log4j/spi/ThrowableInformation.java [hunk 1]
```diff
@@ -29,8 +29,8 @@
 
     static final long serialVersionUID = -4748765566864322735L;
 
-    private transient Throwable throwable;
-    private Method toStringList;
+    private final transient Throwable throwable;
```
[the remark this answers is not in the document]
Comment (turbanoff) [reply log4j-1.2-api/src/main/java/org/apache/log4j/spi/ThrowableInformation.java]: Can you please elaborate? I don't see reasons to not do them `final`

**Comment by remkop** [comment]
As for why transient fields should not be `final`: https://stackoverflow.com/questions/2968876/final-transient-fields-and-serialization

File: log4j-1.2-api/src/main/java/org/apache/log4j/spi/ThrowableInformation.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (rgoers) [reply log4j-1.2-api/src/main/java/org/apache/log4j/spi/ThrowableInformation.java]: final fields are either initialized in the constructor or as assignments on their declaration. Many fields that are declared transient because the values assigned to them are not serializable, not because it is OK for them to be null when the object is reinstantiated via deserialization. It is quite possible that the objects can be recreated during deserialization from the data that is present. But if the fields are marked final those recreated objects cannot be stored in the appropriate fields without using reflection. 

In summary, marking a field as transient simply means it cannot be automatically serialized and deserialized. In no way does it imply that is expected to always be null.

[the remark this answers is not in the document]
Comment (turbanoff) [reply log4j-1.2-api/src/main/java/org/apache/log4j/spi/ThrowableInformation.java]: >It is quite possible that the objects can be recreated during deserialization from the data that is present.

But there is no such code. It means they are not recreated. I someone in future will deserialization code - he can also remove `final` modifier.

Moreover if classes `FormattedMessage`, `StringFormattedMessage` and `ParameterizedMessage`: code is already ready to have this fields `final` and `null`. Such fields are not serialized and not read from ObjectOutputStream.

If you insist - I will remove final

[the remark this answers is not in the document]
Comment (turbanoff) [reply log4j-1.2-api/src/main/java/org/apache/log4j/spi/ThrowableInformation.java]: Just couple examples to show, that having `final transient` it's ok:
https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/time/ZoneRegion.java#L104
https://github.com/OpenGamma/Strata/blob/master/modules/collect/src/main/java/com/opengamma/strata/collect/array/DoubleMatrix.java#L66
https://github.com/google/guava/blob/master/guava/src/com/google/common/reflect/TypeToken.java#L701

File: log4j-api/src/main/java/org/apache/logging/log4j/message/FormattedMessage.java [hunk 2]
```diff
@@ -37,7 +37,7 @@
     private static final Pattern MSG_PATTERN = Pattern.compile(FORMAT_SPECIFIER);
 
     private String messagePattern;
-    private transient Object[] argArray;
+    private final transient Object[] argArray;
```
[the remark this answers is not in the document]
Comment (turbanoff) [reply log4j-api/src/main/java/org/apache/logging/log4j/message/FormattedMessage.java]: fixed

File: log4j-api/src/main/java/org/apache/logging/log4j/message/ParameterizedMessage.java [hunk 3]
```diff
@@ -69,10 +69,10 @@
     private static final int HASHVAL = 31;
 
     // storing JDK classes in ThreadLocals does not cause memory leaks in web apps, so this is okay
-    private static ThreadLocal<StringBuilder> threadLocalStringBuilder = new ThreadLocal<>();
+    private static final ThreadLocal<StringBuilder> threadLocalStringBuilder = new ThreadLocal<>();
 
     private String messagePattern;
-    private transient Object[] argArray;
+    private final transient Object[] argArray;
```
[the remark this answers is not in the document]
Comment (turbanoff) [reply log4j-api/src/main/java/org/apache/logging/log4j/message/ParameterizedMessage.java]: fixed

File: log4j-api/src/main/java/org/apache/logging/log4j/message/StringFormattedMessage.java [hunk 4]
```diff
@@ -44,7 +44,7 @@
     private static final int HASHVAL = 31;
 
     private String messagePattern;
-    private transient Object[] argArray;
+    private final transient Object[] argArray;
```
[the remark this answers is not in the document]
Comment (turbanoff) [reply log4j-api/src/main/java/org/apache/logging/log4j/message/StringFormattedMessage.java]: fixed
